esp-hal/src/i2s/master/camera.rs

```diff
@@ -0,0 +1,404 @@
+        crate::into_mapped_ref!(d4);
+        crate::into_mapped_ref!(d5);
+        crate::into_mapped_ref!(d6);
+        crate::into_mapped_ref!(d7);
+
+        let pairs = [
+            (d0, I2S::din0_signal()),
+            (d1, I2S::din1_signal()),
+            (d2, I2S::din2_signal()),
+            (d3, I2S::din3_signal()),
+            (d4, I2S::din4_signal()),
+            (d5, I2S::din5_signal()),
+            (d6, I2S::din6_signal()),
+            (d7, I2S::din7_signal()),
+        ];
+
+        for (pin, signal) in pairs {
+            pin.init_input(Pull::None);
+            signal.connect_to(pin);
+        }
+
+        self
+    }
+
+    /// Configures the pixel clock pin for the camera interface.
+    pub fn with_ws<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::ws_in_signal().connect_to(pin);
+        self
+    }
+
+    /// Configures the vertical sync (VSYNC) pin for the camera interface.
+    pub fn with_vsync<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::v_sync_signal().connect_to(pin);
+        self
+    }
+
+    /// Configures the horizontal sync (HSYNC) pin for the camera interface.
+    pub fn with_hsync<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::h_sync_signal().connect_to(pin);
+        self
+    }
+
+    /// Configures the vertical sync enable pin for the camera interface.
+    pub fn with_henable<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::h_enable_signal().connect_to(pin);
+        self
+    }
+}
+
+impl<'d, I2S> Camera<'d, I2S>
+where
+    I2S: RegisterAccess,
+{
+    fn start_rx_transfer<'t, RXBUF>(
+        &'t mut self,
+        buf: &'t mut RXBUF,
+        len: usize,
+    ) -> Result<(), Error>
+    where
+        RXBUF: DmaRxBuffer,
+    {
+        if len % 4 != 0 {
+            return Err(Error::IllegalArgument);
+        }
+
+        // Reset RX unit and RX FIFO
+        self._i2s.reset_rx();
+
+        // Enable corresponding interrupts if needed
+
+        // configure DMA outlink
+        unsafe {
+            self.rx_channel
+                .prepare_transfer(self._i2s.dma_peripheral(), buf)
+                .and_then(|_| self.rx_channel.start_transfer())?;
+        }
+
+        // set I2S_RX_STOP_EN if needed
+
+        // start: set I2S_RX_START
+        self._i2s.rx_start(len);
+        Ok(())
+    }
+
+    /// Starts a DMA transfer to receive data from the camera peripheral.
+    pub fn receive<BUF: DmaRxBuffer>(
+        mut self,
+        mut buf: BUF,
+        len: usize,
+    ) -> Result<CameraTransfer<'d, I2S, BUF>, Error> {
+        self.start_rx_transfer(&mut buf, len)?;
```

Comment:
Mind just pasting the body of this function here? No point having a separate function now